token/core/zkatdlog/crypto/rp/bulletproof.go

```diff
@@ -0,0 +1,422 @@
+	}
+	y := v.Curve.HashToZr(bytesToHash)
+	z := v.Curve.HashToZr(y.Bytes())
+	// z^2 and z^3
+	zSquare := z.PowMod(v.Curve.NewZrFromInt(2))
+	zCube := v.Curve.ModMul(zSquare, z, v.Curve.GroupOrder)
+
+	yPow := make([]*math.Zr, len(v.RightGenerators))
+	ipy := v.Curve.NewZrFromInt(0)
+	ip2 := v.Curve.NewZrFromInt(0)
+	for i := 0; i < len(yPow); i++ {
+		// y^i
+		yPow[i] = y.PowMod(v.Curve.NewZrFromInt(int64(i)))
+		// 2^i
+		power2 := v.Curve.NewZrFromInt(2).PowMod(v.Curve.NewZrFromInt(int64(i)))
+		// ipy = \sum y^i
+		ipy = v.Curve.ModAdd(ipy, yPow[i], v.Curve.GroupOrder)
+		// ip2 = sum 2^i
+		ip2 = v.Curve.ModAdd(ip2, power2, v.Curve.GroupOrder)
+	}
+	// polEval = (z -z^)\sum y^i + z^3\sum 2^i
+	polEval := v.Curve.ModSub(z, zSquare, v.Curve.GroupOrder)
+	polEval = v.Curve.ModMul(polEval, ipy, v.Curve.GroupOrder)
+	zCube = v.Curve.ModMul(zCube, ip2, v.Curve.GroupOrder)
+
+	polEval = v.Curve.ModSub(polEval, zCube, v.Curve.GroupOrder)
+
+	// com is should be equal to v.Commitment^{z^2} if p.Value falls within range
+	com := v.CommitmentGenerators[0].Mul(rp.InnerProduct)
+	com.Add(v.CommitmentGenerators[1].Mul(rp.Tau))
+	com.Sub(rp.T1.Mul(x))
+	com.Sub(rp.T2.Mul(xSquare))
+
+	comPrime := v.Commitment.Mul(zSquare)
+	comPrime.Add(v.CommitmentGenerators[0].Mul(polEval))
+
+	if !com.Equals(comPrime) {
+		return errors.New("invalid range proof")
+	}
+
+	// verify the IPA
+	err = v.verifyIPA(rp, x, yPow, z, zSquare)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// preprocess prepares data for the inner product argument
+func (p *rangeProver) preprocess() ([]*math.Zr, []*math.Zr, *math.Zr, *RangeProof, error) {
+	var left, right []*math.Zr
+	var randomLeft, randomRight []*math.Zr
+	rand, err := p.Curve.Rand()
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+	rho := p.Curve.NewRandomZr(rand)
+	eta := p.Curve.NewRandomZr(rand)
+	for i := 0; i < int(p.BitLength); i++ {
+		b := 1 << uint(i) & p.value
+		if b > 0 {
+			b = 1
+		}
+		// this is an array of the bits b_i of p.value
+		left = append(left, p.Curve.NewZrFromInt(int64(b)))
+		// this is an array of b_i - 1
+		right = append(right, p.Curve.ModSub(left[i], p.Curve.NewZrFromInt(1), p.Curve.GroupOrder))
+		// these are randomly generated arrays
+		randomLeft = append(randomLeft, p.Curve.NewRandomZr(rand))
+		randomRight = append(randomRight, p.Curve.NewRandomZr(rand))
+	}
+
+	// C commits to L = (b_0, ..., b_{p.BitLength}) and R = (b_0 - 1 , ..., b_{p.BitLength} - 1)
+	C := commitVector(left, right, p.LeftGenerators, p.RightGenerators, p.Curve)
+	// C is a hiding commitment thanks to rho
+	C.Add(p.P.Mul(rho))
+
+	// D commits two random vectors U and V
+	D := commitVector(randomLeft, randomRight, p.LeftGenerators, p.RightGenerators, p.Curve)
+	// D is a hiding commitment thanks to eta
+	D.Add(p.P.Mul(eta))
+
+	array := common.GetG1Array([]*math.G1{C, D})
+	// compute challenges y and z
+	bytesToHash, err := array.Bytes()
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
```

Comment:
I noticed you don't hash the public parameters here as you did in other places. Is there a reason?

Reply:
There is no need to hash the public parameters.  
However, what's missing here is the commitment to the value which must be added. 
In the branch dedicated to cleaning up the code including the public parameters in the hash will be removed.